src/main/java/com/google/sps/servlets/GenerateServlet.java

```diff
@@ -28,14 +32,143 @@
 
   private static final String FILTER_ARRAY = "clicked-array";
 
+  String[] allPlaces = {"Paris", "New York City", "San Francisco", "London", "Sydney", "Venice"};
+  String[] allDifficulties = {"Easy", "Medium", "Hard"};
+
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Get array of clicked filters and convert to ArrayList<String>
-    Gson gson = new Gson();
     HashSet<String> clickedFilters =
-        gson.fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
+        new Gson().fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
+
+    // Create fake destinations TODO: get destinations from datastore
+    ArrayList<Destination> allDestinations = createFakeDestinations();
+
+    // Create arrays for each category of filter
+    HashSet<String> userPlaces = new HashSet(clickedFilters);
+    HashSet<String> userDifficultyStrings = new HashSet(clickedFilters);
+
+    // Retain only the filters that user chose (for each category)
+    userPlaces.retainAll(Arrays.asList(allPlaces));
+    userDifficultyStrings.retainAll(Arrays.asList(allDifficulties));
+
+    // Convert difficulty level strings to Destination.Obscurity
+    HashSet<Destination.Obscurity> userDifficultyLevels =
+        convertStringsToEnum(userDifficultyStrings);
+
+    // Filter
+    Set<Destination> filteredDestinations =
+        filter(allDestinations, userPlaces, userDifficultyLevels);
+
+    writeToDataStore(filteredDestinations);
 
     response.setContentType("text/html;");
     response.getWriter().println(clickedFilters);
   }
+
+  /* Takes strings of difficulty level and converts into Destination.Obscurity objects. */
+  public HashSet<Destination.Obscurity> convertStringsToEnum(
+      HashSet<String> userDifficultyStrings) {
+    HashSet<Destination.Obscurity> userDifficultyLevels = new HashSet();
+    if (userDifficultyStrings.isEmpty()) {
+      userDifficultyLevels.add(Destination.Obscurity.EASY);
+      userDifficultyLevels.add(Destination.Obscurity.MEDIUM);
+      userDifficultyLevels.add(Destination.Obscurity.HARD);
+    } else {
+      for (String userLevel : userDifficultyStrings) {
+        switch (userLevel) {
+          case "Easy":
+            userDifficultyLevels.add(Destination.Obscurity.EASY);
+            break;
+          case "Medium":
+            userDifficultyLevels.add(Destination.Obscurity.MEDIUM);
+            break;
+          case "Hard":
+            userDifficultyLevels.add(Destination.Obscurity.HARD);
+            break;
+        }
+      }
+    }
+    return userDifficultyLevels;
+  }
```

Comment:
I'm seeing some overlap between this method and the ones that @rysimone created in DestinationDataServlet
https://github.com/googleinterns/step183-2020/blob/master/src/main/java/com/google/sps/servlets/DestinationDataServlet.java#L140

I think it would probably be a good idea to create either a Utils class that could share some of this converter logic or create a static method in the Destination class that could covert a String to a Destination.Obscurity / Destination.Tag enum.

Reply:
I created a static method in the Destination class to do this!